### PR TITLE
Make verify-commits.sh test that merges are clean

### LINKED ISSUE #8089: verify-commits should also check for malicious code in merge commits
Assuming that the machine of a maintainer is compromised, it is possible to have the merge commit modified before sign and push.

As we already have the requirement that "everyone without exception contributes patch proposals using pull requests." [[1]](https://github.com/bitcoin/bitcoin/blob/37f9a1f627c0995d89b62923e75cd092600894f9/CONTRIBUTING.md#contributor-workflow), I'd propose that we add the requirement that a pull request can only be merged when there is no merge conflict to be solved. Thus, verify-commits could not only check the signature of every merge commit but also fail when the merge commit modifies more code than the branch did that is about to be merged.

### PR DESCRIPTION
**Pull request by sipa**
Unsure if we want this.

This modifies verify-commits.sh to redo all merges along the leftmost commit branch (which includes all PR merges), and verify whether they match the merge commit's trees.

The benefit is that it will detect a case where one of the maintainers merges a PR, but makes an unrelated change inside the merge commit. This on itself is not very strong, as unrelated changes can also be included in the merged branch itself - but perhaps the merge commit is not something that people are otherwise likely to look at.

Fixes #8089

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Thanks. Concept ACK

**Comment by donaloconnor** [comment]
Concept Ack

**Comment by laanwj** [comment]
Concept ACK, this makes sense because the merge script (that everyone is supposed to use) will only do clean merges.

- How does this affect performance of the verify script? it's already very slow, so the difference might be negligible)
- How deep should this check be done?

**Comment by sipa** [comment]
The script becomes several times slower (I didn't measure; just my impression from waiting... maybe a few minutes). Perhaps we should restrict it to just one day of commits or so?

**Comment by MarcoFalke** [comment]
Similar to the SHA512 hash, we could only check it for HEAD. (Or maybe make that configurable to check all the way down to trusted_root if you have time)

**Comment by MarcoFalke** [comment]
Fixes #8089

**Review by ryanofsky — COMMENTED** [review]
utACK 577f11141c5155345c68809aac8511b018cc4050, just reviewing the shell script code. Concept also seems good. I'm almost surprised there is no standalone tool like `git-is-clean-merge <commit>`, since merge commits are more cumbersome to verify than normal commits.

**Comment by laanwj** [comment]
utACK 577f111

**Comment by sipa** [comment]
Let's observe how much this adds to Travis checks of master (it does not run on PR branches). If it's unacceptably slow, we can modify the script to only test merge cleanliness for the last few merges or days or so.

File: contrib/verify-commits/verify-commits.sh [hunk 1]
```diff
@@ -123,9 +125,29 @@ while true; do
...
+		PARENTX=${PARENTS#* }
+		PARENT2=${PARENTX%% *}
+		if [ "x$PARENT2" != "x$PARENTX" ]; then
+			echo "Commit $CURRENT_COMMIT is an octopus merge" > /dev/stderr
+			exit 1
+		fi
+	fi
+	if [ "x$PARENT2" != "x" ]; then
+		CURRENT_TREE="$(git show --format="%T" "$CURRENT_COMMIT")"
+		git checkout --force --quiet "$PARENT1"
```
Comment (MarcoFalke) [inline contrib/verify-commits/verify-commits.sh]: Wouldn't a checkout also checkout the other files in `./contrib/verify-commit`, such as `trusted-git-root` or `trusted-keys` and thus change them under its own feet?

**Comment by ken2812221** [comment]
Merge #8672 is not clean, maybe we should ignore this

**Pull request merged**